src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultRazorConfigurationService.cs

```diff
@@ -50,59 +74,91 @@ public async override Task<RazorLSPOptions> GetLatestOptionsAsync(CancellationTo
-                return instance;
+            trace = defaultOptions.Trace;
+            if (razor.TryGetValue("trace", out var parsedTrace))
+            {
+                trace = JTokenToObject(parsedTrace, trace);
             }
-            catch (Exception ex)
+
+            enableFormatting = defaultOptions.EnableFormatting;
+            if (razor.TryGetValue("format", out var parsedFormat))
             {
-                _logger.LogWarning($"Failed to sync client configuration on the server: {ex}");
-                return null;
+                if (parsedFormat is JObject jObject &&
+                    jObject.TryGetValue("enable", out var parsedEnableFormatting))
+                {
+                    enableFormatting = JTokenToObject(parsedEnableFormatting, enableFormatting);
+                }
+            }
+
+            autoClosingTags = defaultOptions.AutoClosingTags;
+            if (html.TryGetValue("autoClosingTags", out var parsedAutoClosingTags))
+            {
+                autoClosingTags = JTokenToObject(parsedAutoClosingTags, autoClosingTags);
             }
         }
 
-        private RazorLSPOptions BuildOptions(IConfiguration config)
+        private static void UpdateVSOptions(
+            RazorLSPOptions defaultOptions,
+            JObject[] result,
+            out bool insertSpaces,
+            out int tabSize)
         {
-            var instance = RazorLSPOptions.Default;
-
-            Enum.TryParse(config["razor:trace"], out Trace trace);
+            var vsEditor = result[2];
 
-            var enableFormatting = instance.EnableFormatting;
-            if (bool.TryParse(config["razor:format:enable"], out var parsedEnableFormatting))
+            insertSpaces = defaultOptions.InsertSpaces;
+            if (vsEditor.TryGetValue(nameof(EditorSettings.IndentWithTabs), out var parsedInsertTabs))
             {
-                enableFormatting = parsedEnableFormatting;
+                insertSpaces = !JTokenToObject(parsedInsertTabs, insertSpaces);
             }
 
-            var autoClosingTags = instance.AutoClosingTags;
-            if (bool.TryParse(config["html:autoClosingTags"], out var parsedAutoClosingTags))
+            tabSize = defaultOptions.TabSize;
+            if (vsEditor.TryGetValue(nameof(EditorSettings.IndentSize), out var parsedTabSize))
             {
-                autoClosingTags = parsedAutoClosingTags;
+                tabSize = JTokenToObject(parsedTabSize, tabSize);
             }
+        }
 
-            return new RazorLSPOptions(trace, enableFormatting, autoClosingTags);
+        private static T JTokenToObject<T>(JToken token, T defaultValue)
+        {
+            try
+            {
+                // JToken.ToObject could potentially throw here if the user provides malformed options.
+                // If this occurs, catch the exception and return the default value.
+                return token.ToObject<T>();
```

Comment:
I forget if it's possible for this to return null. If so, would:

```suggestion
                return token.ToObject<T>() ?? defaultValue;
```

be preferable?